core/src/HIP/Kokkos_HIP_Parallel_Range.hpp

```diff
@@ -507,15 +501,17 @@ class ParallelScanHIPBase {
   //----------------------------------------
 
   __device__ inline void initial() const {
-    const integral_nonzero_constant<size_type, ValueTraits::StaticValueSize /
+    typename Analysis::Reducer final_reducer(&m_functor);
+
+    const integral_nonzero_constant<size_type, Analysis::StaticValueSize /
                                                    sizeof(size_type)>
-        word_count(ValueTraits::value_size(m_functor) / sizeof(size_type));
+        word_count(Analysis::value_size(m_functor) / sizeof(size_type));
 
     size_type* const shared_value =
         Kokkos::Experimental::kokkos_impl_hip_shared_memory<size_type>() +
         word_count.value * threadIdx.y;
 
-    ValueInit::init(m_functor, shared_value);
+    final_reducer.init(reinterpret_cast<pointer_type>(shared_value));
```

Comment:
The two times you use `shared_value`, you need to do a `reinterpret_cast`. Change the type of `shared_value` instead.